File: LeetCodeAlgos/Python/mergeTwoSortedLists.py
# You are given the heads of two sorted linked lists list1 and list2.
# Merge the two lists in a one sorted list. The list should be made by splicing together the nodes of the first two lists.
# Return the head of the merged linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def mergeTwoLists(self, list1, list2):
        res = newNode =  ListNode()
        while list1 and list2:
            if list1.val <= list2.val:
                newNode.next = list1
                list1 = list1.next
                newNode = newNode.next
            else:
                newNode.next = list2
                list2 = list2.next
                newNode = newNode.next
        if list1:
            newNode.next = list1
        if list2:
            newNode.next = list2
        return res.next

s = Solution()
print(s.mergeTwoLists(ListNode(1, ListNode(2, ListNode(4))),ListNode(1, ListNode(3, ListNode(5)))))
print(s.mergeTwoLists([],[]))
print(s.mergeTwoLists([],[ListNode(0)]))

File: LeetCodeAlgos/Java/MergeTwoSortedLists.java
// You are given the heads of two sorted linked lists list1 and list2.

// Merge the two lists into one sorted list. The list should be made by splicing together the nodes of the first two lists.

// Return the head of the merged linked list.

public class MergeTwoSortedLists {
  
}

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
  public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
      ListNode ans = new ListNode(Integer.MIN_VALUE);
      ListNode mutator = ans;
      while (list1 != null && list2 != null) {
        if (list1.val <= list2.val) {
          mutator.next = list1;
          list1 = list1.next;
        } else {
          mutator.next = list2;
          list2 = list2.next;
        }
        mutator = mutator.next;
        mutator.next = null;
      }
      mutator.next = list1 == null ? list2 : list1;
      return ans.next;
  }
}